File: src/main/java/leetcode/editor/cn/TypeDefined.java
package leetcode.editor.cn;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * scm.com Inc.
 * Copyright (c) 2004-2021 All Rights Reserved.
 */
public class TypeDefined {

    public static class ListNode {
        public int val;
        public ListNode next;

        public ListNode() {
        }

        public ListNode(int val) {
            this.val = val;
        }

        public ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            ListNode current = this;
            boolean isFirst = true;
            while (current != null) {
                if (isFirst) {
                    isFirst = false;
                } else {
                    sb.append("->");
                }
                sb.append(current.val);
                current = current.next;
            }
            return sb.toString();
        }

        public static ListNode build(int... values) {
            ListNode h = new ListNode(-1);
            ListNode p = h;
            for (int value : values) {
                p.next = new ListNode(value);
                p = p.next;
            }
            return h.next;
        }
    }


    /**
     * 整型列表
     *
     * @param arr
     * @return
     */
    public static List<Integer> toList(int[] arr) {
        return Arrays.stream(arr).mapToObj(i -> i).collect(Collectors.toList());
    }

    /**
     * 整型数组
     *
     * @param list
     * @return
     */
    public static int[] toInts(List<Integer> list) {
        int[] arr = new int[list.size()];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = list.get(i);
        }
        return arr;
    }

}

File: src/main/scala/com/scorpio/leetcode/editor/cn/TypeDefined.scala
package com.scorpio.leetcode.editor.cn

/**
 * scm.com Inc.
 * Copyright (c) 2004-2021 All Rights Reserved.
 */
object TypeDefined {
  class ListNode(_x: Int = 0, _next: ListNode = null) {
    var x = _x;
    var next = _next

    /**
     * 打印链表
     *
     * @return
     */
    override def toString: String = {
      val sb = new StringBuilder
      var current = this
      var isFirst = true
      while (current != null) {
        if (isFirst) isFirst = false
        else sb.append("->")
        sb.append(current.x)
        current = current.next
      }
      sb.toString
    }
  }

  object ListNode {
    /**
     * 构建链表
     *
     * @param values
     * @return
     */
    def build(values: Int*): TypeDefined.ListNode = {
      val h = new TypeDefined.ListNode(-1)
      var p = h
      for (value <- values) {
        p.next = new TypeDefined.ListNode(value)
        p = p.next
      }
      h.next
    }
  }
}
